crates/sui-framework/packages/sui-system/sources/validator_set.move

```diff
@@ -233,11 +238,40 @@ public(package) fun request_add_validator(self: &mut ValidatorSet, min_joining_s
         EDuplicateValidator
     );
     assert!(validator.is_preactive(), EValidatorNotCandidate);
-    assert!(validator.total_stake_amount() >= min_joining_stake_amount, EMinJoiningStakeNotReached);
+    assert!(self.can_join(validator.total_stake_amount(), ctx), EMinJoiningStakeNotReached);
 
     self.pending_active_validators.push_back(validator);
 }
 
+/// Return `true` if a  candidate validator with `stake` will have sufficeint voting power to join the validator set
+fun can_join(self: &ValidatorSet, stake: u64, ctx: &TxContext): bool {
+    let (min_joining_voting_power, _, _) = self.get_voting_power_thresholds(ctx);
+    // if the validator will have at least `min_joining_voting_power` after joining, they can join.
+    // this formula comes from SIP-39 TODO link once landed
+    let future_total_stake = self.total_stake + stake;
+    let future_validator_voting_power = voting_power::derive_voting_power(stake, future_total_stake);
+    future_validator_voting_power >= min_joining_voting_power
+}
+
+/// return (min, low, very low voting power) thresholds
+fun get_voting_power_thresholds(self: &ValidatorSet, ctx: &TxContext): (u64, u64, u64) {
+    let start_epoch = {
+        let key = VotingPowerAdmissionStartEpochKey();
+        if (self.extra_fields.contains(key)) *self.extra_fields.borrow(key)
+        else ctx.epoch() + 1 // will give us the phase 1 values
+    };
+    // these numbers come from SIP-39 TODO link once landed
+    let phase_length = 14; // phases are 14 days = 14 epochs
+    let epoch = ctx.epoch();
+    if (epoch < start_epoch + phase_length) { // phase 1
+        (12, 8, 4)
+    } else if (epoch < start_epoch + (2 * phase_length)) { // phase 2
+        (6, 4, 2)
+    } else { // phase 3
+        (3, 2, 1)
+    }
```

Comment:
Nit, do we want `{}` on the if-else bodies here?

Reply:
I suggest we keep it as is, and patch later with other formatting changes. Also, discovered a bug in the formatter related to line comments, which I'll have to fix to make that happen.